src/Features/Core/Portable/Diagnostics/Analyzers/PreferFrameworkTypeDiagnosticAnalyzerBase.cs

```diff
@@ -0,0 +1,114 @@
+                DiagnosticSeverity.Hidden,
+                isEnabledByDefault: true);
+
+        public bool RunInProcess => true;
+
+        public DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            s_descriptorPreferFrameworkTypeInDeclarations, s_descriptorPreferFrameworkTypeInMemberAccess);
+
+        protected abstract ImmutableArray<TSyntaxKind> SyntaxKindsOfInterest { get; }
+        protected abstract bool IsPredefinedTypeReplaceableWithFrameworkType(TPredefinedTypeSyntax node);
+        protected abstract bool IsInMemberAccessOrCrefReferenceContext(TExpressionSyntax node);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKindsOfInterest);
+        }
+
+        protected void AnalyzeNode(SyntaxNodeAnalysisContext context)
+        {
+            var optionSet = context.Options.GetOptionSet();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var predefinedTypeNode = (TPredefinedTypeSyntax)context.Node;
+
+            // check if the predefined type is replaceable with an equivalent framework type.
+            if (!IsPredefinedTypeReplaceableWithFrameworkType(predefinedTypeNode))
+            {
+                return;
+            }
+
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+
+            // check we have a symbol so that the fixer can generate the right type syntax from it.
+            var typeSymbol = semanticModel.GetSymbolInfo(predefinedTypeNode, cancellationToken).Symbol as ITypeSymbol;
+            if (typeSymbol == null)
+            {
+                return;
+            }
+
+            string applicableDiagnosticId;
+            PerLanguageOption<CodeStyleOption<bool>> applicableOption;
+
+            // we have a predefined type syntax that is either in a member access context or a declaration context. 
+            if (IsInMemberAccessOrCrefReferenceContext(predefinedTypeNode))
+            {
+                applicableOption = CodeStyleOptions.PreferIntrinsicPredefinedTypeKeywordInMemberAccess;
+                applicableDiagnosticId = IDEDiagnosticIds.PreferFrameworkTypeInMemberAccessDiagnosticId;
+            }
+            else
+            {
+                applicableOption = CodeStyleOptions.PreferIntrinsicPredefinedTypeKeywordInDeclaration;
+                applicableDiagnosticId = IDEDiagnosticIds.PreferFrameworkTypeInDeclarationsDiagnosticId;
+            }
+
+            var optionValue = optionSet.GetOption(applicableOption, semanticModel.Language);
+            var preferFrameworkType = !optionValue.Value;
+
+            if (preferFrameworkType && optionValue.Notification.Value != DiagnosticSeverity.Hidden)
```

Comment:
checking both options and insta-bailing is really not that bad.  especially as there will be people that do not want any style checking, and it would be nice to not cost them anything.